WordPressAuthenticator/Email Client Picker/AppSelector.swift

```diff
@@ -0,0 +1,122 @@
+                return nil
+            }
+            //sort the apps alphabetically
+            actions = actions.sorted { $0.title ?? "" < $1.title ?? "" }
+            actions.append(UIAlertAction(title: AppSelectorTitles.cancel.localized, style: .cancel, handler: nil))
+
+            if let action = defaultAction {
+                actions.insert(action, at: 0)
+            }
+            return actions
+        }
+
+        guard let appCalls = buildAppCalls(from: appList) else {
+            return nil
+        }
+
+        alertController = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        appCalls.forEach {
+            alertController.addAction($0)
+        }
+    }
+}
+
+
+/// Initializers for Email Picker
+extension AppSelector {
+    /// initializes the picker with a plist file in a specified bundle
+    convenience init?(with plistFile: String, in bundle: Bundle, defaultAction: UIAlertAction? = nil) {
+
+        guard let plistPath = bundle.path(forResource: plistFile, ofType: "plist"),
+            let availableApps = NSDictionary(contentsOfFile: plistPath) as? [String: String] else {
+            return nil
+        }
+        self.init(with: availableApps, defaultAction: defaultAction)
+    }
+
+    /// Convenience init for a picker that calls supported email clients apps, defined in EmailClients.plist
+    convenience init?() {
+        guard let bundlePath = Bundle(for: type(of: self))
+            .path(forResource: "WordPressAuthenticatorResources", ofType: "bundle"),
+            let wpAuthenticatorBundle = Bundle(path: bundlePath) else {
+                return nil
+        }
+
+        let plistFile = "EmailClients"
+        var defaultAction: UIAlertAction?
+
+        // if available, prepend apple mail
+        if MFMailComposeViewController.canSendMail(), let url = URL(string: "message://") {
+            defaultAction = UIAlertAction(title: AppSelectorTitles.appleMail.localized, style: .default) { action in
+                UIApplication.shared.open(url)
+            }
+        }
+        self.init(with: plistFile, in: wpAuthenticatorBundle, defaultAction: defaultAction)
+    }
+}
+
+
+/// Localizable app selector titles
+enum AppSelectorTitles: String {
+    case appleMail
+    case gmail
+    case airmail
+    case msOutlook
+    case spark
+    case yahooMail
+    case fastmail
+    case cancel
+
+    var localized: String {
+        switch self {
+        case .appleMail:
+            return NSLocalizedString("Mail (Default)", comment: "Label for Apple Mail as default email client")
+        case .gmail:
+            return NSLocalizedString("Gmail", comment: "Gmail")
+        case .airmail:
+            return NSLocalizedString("Airmail", comment: "Airmail")
+        case .msOutlook:
+            return NSLocalizedString("Microsoft Outlook", comment: "Microsoft Outlook")
+        case .spark:
+            return NSLocalizedString("Spark", comment: "Spark")
```

Comment:
Do you think most of these need to be localised? I may be wrong, but I think e.g. Outlook and Gmail are still called the same thing in other languages? If we do keep these as localized strings, we may want to provide a little more context in the comments – for example, `Spark` on its own doesn't indicate that it's the name of an email client. Perhaps `Option to select the Spark email client when logging in with magic links`?

Reply:
I thought about that, I did it mostly for languages that have different character sets than English. I will provide more context.